src/MessagePack/Internal/UnsafeRefSerializeHelper.cs

```diff
@@ -0,0 +1,1572 @@
+﻿// Copyright (c) All contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Buffers.Binary;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+#if NET8_0_OR_GREATER
+using System.Runtime.Intrinsics;
+#endif
+
+namespace MessagePack.Internal;
+
+internal static class UnsafeRefSerializeHelper
+{
+    internal static void Serialize(ref MessagePackWriter writer, ref bool input, int length)
+    {
+#if NET8_0_OR_GREATER
+        unchecked
+        {
+            if (Vector128.IsHardwareAccelerated)
+            {
+                const int mask = ~15;
+                var alignedInputLength = (nuint)(uint)(length & mask);
+                if (alignedInputLength > 0)
+                {
+                    var destination = writer.GetSpan((int)alignedInputLength);
+                    ref var outputIterator = ref MemoryMarshal.GetReference(destination);
+                    for (nuint inputOffset = 0; inputOffset < alignedInputLength; inputOffset += (nuint)Vector128<byte>.Count)
+                    {
+                        writer.CancellationToken.ThrowIfCancellationRequested();
+
+                        // Load 16 bools.
+                        var loaded = Vector128.LoadUnsafe(ref Unsafe.As<bool, sbyte>(ref input), inputOffset);
+
+                        // true is not 0, but false is always 0 in C#. true -> 0. false -> -1.
+                        var falses = Vector128.Equals(loaded, Vector128<sbyte>.Zero);
+
+                        // MessagePackCode.True is 0xc3(-61). MessagePackCode.False is 0xc2(-62).
+                        var results = Vector128.Create(unchecked((sbyte)MessagePackCode.True)) + falses;
+
+                        // Store 16 values.
+                        results.AsByte().StoreUnsafe(ref outputIterator, inputOffset);
+                    }
+
+                    writer.Advance((int)alignedInputLength);
+                    length -= (int)alignedInputLength;
+                    input = ref Unsafe.Add(ref input, alignedInputLength);
+                }
+            }
+        }
+#endif
```

Comment:
I see no `else` or `return` above this point, so it looks to me like `NET 8` with hardware acceleration will serialize _twice_.

Reply:
No, it does not serialize twice.

I will use `bool` as an example and the environment can use 128bit simd intrinsics.
If input `length` is less than 16, the running code skips the `#if NET8_0_OR_GREATER` block.
The `#if NET8_0_OR_GREATER` block consumes multiples of 16 input items.
Then the following code consumes the remainder of input items.

Reply:
I still don't understand how this doesn't serialize twice. Here is the high level as I interpret it:

1. If NET8 + accelerated, serialize in a fast way.
2. Serialize in a slower way.

The slow route will only execute 2. But the fast route will execute 1, then continue and execute 2 as well. I would expect the fast route to include a `return;` statement to prevent continuing to the slow route.

Reply:
For the sake of clarity, let us assume that the length of the input element is a concrete number, 28.
In this case, first 16 elements are serialized in the faster way.
The remaining 12 elements are serialized in the slower way.
The first 16 elements are never serialized twice.

The faster way cannot handle a number of elements that are not multiples of 16 in the way I have written this time.
The remainder must always be handled in the slower way.

Reply:
Oh! Because you're mutating the `input` reference and decrementing `length` as you write. Got it. Thanks.